### ISSUE TITLE
Re-think initializers

### ISSUE THREAD

**Issue by aidan-casey** [tempestphp/tempest-framework#183]
One thing that has always bothered me about the `#[InitializedBy]` attribute is that it makes the interface somewhat aware of the implementation. I am proposing the following as a solution to this problem. I believe this is a serious issue for Tempest and would love to help drive a solution to that.

Below are some of my initial thoughts here, though I am sure I am missing some edge cases and things you have thought about @brendt, so feel free to school me. 🙂 

# Make Interfaces Unaware of Implementations

The first step to making interfaces unaware of implementations is moving the logic that tells the container how to create an object away from the interface. The initializer itself will inform the container what it provides and, for now, the container will register these in the order they are scanned. In the long run we might add a priority property or something.

To do this, I propose that we replace the `#[InitializedBy()]` with the `#[Initializes()]` attribute which is only available on _methods_.

For example, we might have the following initializer class for our http client:

```php
final class TempestHttpClientInitializer
{
    // Shorthand that wraps this initializer call and makes
    // the result a singleton in the container.
    #[Singleton]

    // Informs the container that this initializer handles
    // this class.
    #[Initializes(HttpClientDriver::class)]

    // Gets called when it's time to initialize HttpClientDriver
    public function initializePsr18Driver(): HttpClientDriver
    {
        return new Psr18Driver(...);
    }

    #[Initializes(HttpClient::class)]
    public function initializeHttpClient(): HttpClient
    {
        return new GenericHttpClient(...);
    }
}
```

Behind the scenes, this basically becomes a register call. The initializer method will be wrapped by a `Callable` class that allows for a reflection method _or_ function to be passed. When it comes time to initialize, we just grab the key for that class and run the `call` method. So something like this:

```php
if (isset($this->singletons[$class]) !== false) {
    return $this->singletons[$class];
}

if (isset($this->definitions[$class]) !== false) {
    // This will actually autowire the initializer,
    // which means we can add our dependencies
    // to the construct if we need to. That said,
    // this may be an initializer, or it may be
    // a callable.
    return $this->definitions[$class]->call();
}

return $this->autowire($class);
```

This has the bonus side effect of enforcing separation of concerns and some good contracts/dependencies.

**Comment by brendt** [comment]
A couple of preliminary thoughts:

- Some initializers initialize multiple classes. Though that shouldn't be a problem since we already have a solution for that in place with `CanInitialize`.
- We could simply look at the return type of the initializer method, eliminating the need for an attribute. It would mean one initializer per class, while in your example I think you want to be able to add multiple methods.
- If we look at the return type, we could add support for union return types
- Whether we use `#[Initializes(HttpClientDriver::class)]` or `function initialize(): HttpClientDriver {}`, I think it's important to note that the class here should be the interface, no? Or did you want to auto-detect the interface based on the concrete class? I think that's a bit too much magic, especially since classes can implement multiple interfaces, and I don't want to deal with managing collisions.

Overall, I like the idea 👍

**Comment by brendt** [comment]
Oh and `#[Singleton]` is a good idea as well 👍

**Comment by brendt** [comment]
I'm working on this :)

**Comment by brendt** [comment]
PR is here: #184

**Comment by aidan-casey** [comment]
> Some initializers initialize multiple classes. Though that shouldn't be a problem since we already have a solution for that in place with CanInitialize.

So I actually have a couple questions on this...

1. When an initializer initializes multiple classes, is this just for dependencies? Keep in mind that we can only return _one_ object today.

2. I _think_ `CanInitialize` can go away with this change. The `Initializes` attribute would be telling the container up front what it handles.

> We could simply look at the return type of the initializer method, eliminating the need for an attribute. It would mean one initializer per class, while in your example I think you want to be able to add multiple methods.

Three problems I foresee with this:

1. We cannot eliminate `CanInitialize` by using the return type. We might initialize `HttpClient` and `GenericHttpClient`. I recognize of course, `CanInitialize` is heavily logic based today and we'd need to have a solution to that.

2. You are discovering against an interface which feels less Tempesty.  

3. I think there are a lot of times that we are initializing multiple dependencies at the same time that really could be moved to their own initializer, but due to the fact that means and entire other class, we shortcut and don't. I think this gives us the ability to not only adhere to that "best practice" but also use those sibling initializers.

Looking at my example again, it might look like this:

```php

final class TempestHttpClientInitializer
{
    // Shorthand that wraps this initializer call and makes
    // the result a singleton in the container.
    #[Singleton]

    // Informs the container that this initializer handles
    // this class.
    #[Initializes(HttpClientDriver::class)]

    // Gets called when it's time to initialize HttpClientDriver
    public function initializePsr18Driver(): HttpClientDriver
    {
        return new Psr18Driver(...);
    }

    #[Initializes(HttpClient::class)]
    public function initializeHttpClient(): HttpClient
    {
        return new GenericHttpClient(
            $this->initializePsr18Driver()
        );
    }
}
```

> I think it's important to note that the class here should be the interface, no?

I don't think it matters, does it? Ideally it'd be an interface, but you might also have:

```php

final class TempestHttpClientInitializer
{
    #[Initializes(HttpClient::class)]

    #[Initializes(GenericHttpClient::class)]

    public function initializeHttpClient(): HttpClient
    {
        return new GenericHttpClient(...);
    }
}
```

**Comment by brendt** [comment]
> We cannot eliminate CanInitialize by using the return type. We might initialize HttpClient and GenericHttpClient. I recognize of course, CanInitialize is heavily logic based today and we'd need to have a solution to that.

I actually think we can, by supporting union types (which I already do in my PR). The only problem I still foresee is with the `RequestInitializer`. Gonna do some tinkering

> You are discovering against an interface which feels less Tempesty.

We actually do this in `DiscoveryDiscovery` and `MigrationDiscovery` as well. So that's 3 out of 7. I actually think relying on the Type system is more Tempesty, since we're doing that for Command and Event handler matching as well.

> I think there are a lot of times that we are initializing multiple dependencies at the same time that really could be moved to their own initializer, but due to the fact that means and entire other class, we shortcut and don't. I think this gives us the ability to not only adhere to that "best practice" but also use those sibling initializers.

Might be true. I'm gonna experiment and get back to you

**Comment by aidan-casey** [comment]
> I actually think we can, by supporting union types (which I already do in my PR). The only problem I still foresee is with the RequestInitializer. Gonna do some tinkering

With union types that makes sense.

RE:Tinkering, can you pass a closure to an attribute? I'm thinking the answer is yes, which might also let us put that logic right there.

> We actually do this in DiscoveryDiscovery and MigrationDiscovery as well. So that's 3 out of 7. I actually think relying on the Type system is more Tempesty, since we're doing that for Command and Event handler matching as well.

Fair enough. 🙂 

> Might be true. I'm gonna experiment and get back to you

Okie dokie!

**Comment by brendt** [comment]
> I actually think we can, by supporting union types (which I already do in my PR). The only problem I still foresee is with the RequestInitializer. Gonna do some tinkering

I managed to remove `CanInitialize` and `RequiresClassName`. I did add a new interface: `DynamicInitializer`. These are initializers that require more than a type check to determine whether they can be used to initialize a certain dependency or not. There are two scenarios where that happens:

- When a concrete `Request` implementation is requested
- During route binding, where a specific model implementation is requested

We might be able to get rid of this interface as well if we can move the `is_a` check to the container. Gonna look into that next.
